wpimath/src/main/native/include/units/base.h

```diff
@@ -3426,6 +3456,76 @@ namespace units {
 			return (lhs > r ? lhs : r);
 		}
 	}
+
+#if __has_include(<wpi/telemetry/TelemetryTable.h>) && !defined(UNIT_LIB_DISABLE_TELEMETRY)
+namespace detail {
+template <typename Ratio, typename Char, typename Traits, size_t N, size_t N1>
+consteval auto ConcatAbbrev(
+		wpi::ct_string<Char, Traits, N> const& str,
+		wpi::ct_string<Char, Traits, N1> const& abbrev) {
+  using namespace wpi::literals;
+	constexpr auto num_str = wpi::NumToCtString<Ratio::num>();
+	constexpr auto den_str = wpi::NumToCtString<Ratio::den>();
+	constexpr int asize =
+		(Ratio::num != 0 ? N1 : 0) +
+		(Ratio::num != 0 && Ratio::num != 1 ? (1 + num_str.size()) : 0) +
+		(Ratio::den != 1 ? (1 + den_str.size()) : 0);
+	constexpr int size = N + (N != 0 && asize != 0 ? 1 : 0) + asize;
+
+	// Need a dummy array to instantiate a ct_string.
+	constexpr Char dummy[1] = {};
+	auto res = wpi::ct_string<Char, Traits, size>{dummy};
+
+	auto p = res.chars.begin();
+  auto append = [&p](auto&& s) {
+    for (auto c : s) {
+      *p++ = c;
+    }
+  };
+
+	append(str);
+	if constexpr (N != 0 && asize != 0) {
+		append(" "_ct_string);
+	}
+	if constexpr (Ratio::num != 0) {
+		append(abbrev);
+	}
+	if constexpr (Ratio::num != 0 && Ratio::num != 1) {
+		append("^"_ct_string);
+		append(num_str);
+	}
+	if constexpr (Ratio::den != 1) {
+		append("/"_ct_string);
+		append(den_str);
+	}
+	return res;
```

Comment:
```suggestion
  constexpr auto only_if = []<bool b>(const auto& s) {
    if constexpr (b) {
      return s;
    }
    return ""_ct_string;
  };
  return wpi::Concat(
    str,
    only_if<N != 0 && Ratio::num != 0>(" "_ct_string),
    only_if<Ratio::num != 0>(abbrev),
    only_if<Ratio::num != 0 && Ratio::num != 1>("^"_ct_string),
    only_if<Ratio::num != 0 && Ratio::num != 1>(wpi::NumToCtString<Ratio::num>()),
    only_if<Ratio::den != 1>("/"_ct_string),
    only_if<Ratio::den != 1>(wpi::NumToCtString<Ratio::den>()));
```
This is a lot clearer about the intent while maintaining the same number of intermediate copies.

Reply:
Clever!  I wanted to do ternaries but of course that doesn't work with different ct_string lengths, didn't think about wrapping it in a constexpr function.